### PR TITLE
PrefetchingIter & upsampling bug fix & L1 Loss

### PR DESCRIPTION
**Pull request by piiswrong**

### CODE REVIEW

**Comment by tqchen** [comment]
The prefetch_next seems to be a bit redundant. Normally, we can create a prefetchIter that behaves like iter, but takes an underlying iter and call source.next to get data.

Also due to python GIL, it might not be useful to create multi-threaded prefetcher in python. The other changes LGTM, can you first make another PR with the other changes?

**Comment by piiswrong** [comment]
It works as long as the other threads are calling c api. Subclassing seems better because it let you expose additional interface of the underlying iter.

**Comment by tqchen** [comment]
- If the data comes from C API, then it is already prefetched by default
- If the purpose is prefetching only, a wrapping class is simpler as it do not introduce new interfaces, which could be hard to maintain, and the prefetch logic is independent from data loading. Something like

```
data = Prefetch(MNISTIter())
```

**Comment by piiswrong** [comment]
For example in your prefected iter you are calling numpy, which is calling cblas, or opencv::resize

**Comment by piiswrong** [comment]
@tqchen refactored. please have a look

File: python/mxnet/metric.py [hunk 1]
```diff
@@ -59,6 +59,17 @@ def update(self, labels, preds):
             num_inst = pred_label.shape[0]
         self.num_inst += num_inst
 
+class L1(EvalMetric):
```
Comment (tqchen) [inline python/mxnet/metric.py]: Use more explicit naming, say L1 Norm

Comment (tqchen) [inline python/mxnet/metric.py]: It is actually Mean Absolute error, MAE from what you did

File: python/mxnet/io.py [hunk 2]
```diff
@@ -92,8 +97,120 @@ def getpad(self):
...
+    prefetching.
 
-DataBatch = namedtuple('DataBatch', ['data', 'label', 'pad', 'index'])
+    Parameters
+    ----------
+    iters : DataIter or list of DataIter
+        one or more DataIters (or any class with "reset" and "read" methods)
+    rename_data : None or list of dict
+        i-th element is a renaming map for i-th iter, in the form of
+        {'original_name' : 'new_name'}. Should have one entry for each entry
```
Comment (tqchen) [inline python/mxnet/io.py]: consider add an example code here

**Pull request merged**